torch/nn/modules/transformer.py

```diff
@@ -85,7 +85,8 @@ def __init__(self, d_model: int = 512, nhead: int = 8, num_encoder_layers: int =
 
     def forward(self, src: Tensor, tgt: Tensor, src_mask: Optional[Tensor] = None, tgt_mask: Optional[Tensor] = None,
                 memory_mask: Optional[Tensor] = None, src_key_padding_mask: Optional[Tensor] = None,
-                tgt_key_padding_mask: Optional[Tensor] = None, memory_key_padding_mask: Optional[Tensor] = None) -> Tensor:
+                tgt_key_padding_mask: Optional[Tensor] = None, memory_key_padding_mask: Optional[Tensor] = None,
+                src_is_causal: bool = False, tgt_is_causal: bool = False, memory_is_causal: bool = False) -> Tensor:
```

Comment:
Should this be Optional[bool] = None ?
and add detection of causal mask similar to https://github.com/pytorch/pytorch/blob/master/torch/nn/modules/transformer.py#L302 in TransformerEncoder
to TransformerDecoder and Transformer as well?
(We can't do thi at the layer level because it's prohibitively expensive, but we can "buy back" the expense of detecting a causal mask if we get a substantial use out of it.  Users can prevent running this code by setting to True/False.



Also, should we replace https://github.com/pytorch/pytorch/blob/master/torch/nn/modules/transformer.py#L305
with a call to Transformer. generate_square_subsequent_mask
(and maybe give that a dtype argument that defaults to torch.float, too?

Reply:
Agree with everything you said; the un-avoidable causal mask detection and not using `genarate_square_subsequent_mask` had bothered me from the PyTorch 2.0 release as well. :)
Also the missing `dtype` argument had always been a thorn.

That's why I love the recent change of not _always_ trying to detect a causal mask by making `is_causal` `Optional`. I hadn't noticed this change while making the PR, but will definitely integrate your suggestions. Thanks!

Is it fine if I just put everything into this PR or would you like me to make different ones (e.g. for the `generate_square_subsequent_mask` `dtype` argument)?

Reply:
I’m good with just amending this the current PR to include these functions